### PR TITLE
Support Keras grouped convolutions

### LINKED ISSUE #36431: TensorFlow2.x should support group convolutions in keras api level.
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.1.0
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.**
Group convolution has been widely used in various deep learning models, such as ResNeXt(https://arxiv.org/abs/1611.05431). PyTorch has already supported group convolution, while TensorFlow has not. As this issue(https://github.com/tensorflow/tensorflow/issues/34024) says, ```tf.nn.conv2d() ``` supports group convolution on GPU, but not on CPU. Therefore, A keras layer api which supports group convolution is needed for many users. This is a sample code which supports the feature by modifying the keras ```Conv2D``` api:
```
class Conv2D(tf.keras.layers.Layer):
    def __init__(self,
                 input_channels,
                 output_channels,
                 kernel_size,
                 strides=(1, 1),
                 padding='valid',
                 data_format=None,
                 dilation_rate=(1, 1),
                 activation=None,
                 groups=1,
                 use_bias=True,
                 kernel_initializer='glorot_uniform',
                 bias_initializer='zeros',
                 kernel_regularizer=None,
                 bias_regularizer=None,
                 activity_regularizer=None,
                 kernel_constraint=None,
                 bias_constraint=None,
                 **kwargs):
        super(Conv2D, self).__init__()

        if not input_channels % groups == 0:
            raise ValueError("The value of input_channels must be divisible by the value of groups.")
        if not output_channels % groups == 0:
            raise ValueError("The value of output_channels must be divisible by the value of groups.")

        self.kernel_size = kernel_size
        self.strides = strides
        self.padding = padding
        self.data_format = data_format
        self.dilation_rate = dilation_rate
        self.activation = activation
        self.groups = groups
        self.use_bias = use_bias
        self.kernel_initializer = kernel_initializer
        self.bias_initializer = bias_initializer
        self.kernel_regularizer = kernel_regularizer
        self.bias_regularizer = bias_regularizer
        self.activity_regularizer = activity_regularizer
        self.kernel_constraint = kernel_constraint
        self.bias_constraint = bias_constraint

        self.group_in_num = input_channels // groups
        self.group_out_num = output_channels // groups
        self.conv_list = []
        for i in range(self.groups):
            self.conv_list.append(tf.keras.layers.Conv2D(filters=self.group_out_num,
                                                         kernel_size=kernel_size,
                                                         strides=strides,
                                                         padding=padding,
                                                         data_format=data_format,
                                                         dilation_rate=dilation_rate,
                                                         activation=activations.get(activation),
                                                         use_bias=use_bias,
                                                         kernel_initializer=initializers.get(kernel_initializer),
                                                         bias_initializer=initializers.get(bias_initializer),
                                                         kernel_regularizer=regularizers.get(kernel_regularizer),
                                                         bias_regularizer=regularizers.get(bias_regularizer),
                                                         activity_regularizer=regularizers.get(activity_regularizer),
                                                         kernel_constraint=constraints.get(kernel_constraint),
                                                         bias_constraint=constraints.get(bias_constraint),
                                                         **kwargs))

    def call(self, inputs, **kwargs):
        feature_map_list = []
        for i in range(self.groups):
            x_i = self.conv_list[i](inputs[:, :, :, i*self.group_in_num: (i + 1) * self.group_in_num])
            feature_map_list.append(x_i)
        out = tf.concat(feature_map_list, axis=-1)
        return out
```

**Will this change the current api? How?**
This will change the current keras convolution api . 

**Who will benefit with this feature?**
Anyone who need to use group conolution in their code will benefit.

**Any Other info.**

### PR DESCRIPTION
**Pull request by lgeiger**
This PR adds support for grouped convolutions to Keras layers.

For a more in-depth discussion see https://github.com/keras-team/governance/pull/16

Closes #36431

### CODE REVIEW

**Comment by lgeiger** [comment]
@gbaned @fchollet Any updates on this?

File: tensorflow/python/keras/layers/convolutional.py
Comment (fchollet) [inline tensorflow/python/keras/layers/convolutional.py]: "The number of filters must be evenly divisible by the number of groups."

File: tensorflow/python/keras/layers/convolutional.py
Comment (fchollet) [inline tensorflow/python/keras/layers/convolutional.py]: "received: groups={}, filters={}"

File: tensorflow/python/keras/layers/convolutional.py
Comment (fchollet) [inline tensorflow/python/keras/layers/convolutional.py]: `if filters` seems superfluous since filters=0 is invalid

File: tensorflow/python/keras/layers/convolutional.py
Comment (fchollet) [inline tensorflow/python/keras/layers/convolutional.py]: "The number of input channels must be evenly divisible by the number of groups. Received groups={}, but the input has {} channels (full input shape is {})"

File: tensorflow/python/ops/nn_ops.py
Comment (fchollet) [inline tensorflow/python/ops/nn_ops.py]: number of output filters

Comment (fchollet) [inline tensorflow/python/ops/nn_ops.py]: provide the values of the passed argument, with their names

File: tensorflow/python/keras/layers/convolutional_test.py
Comment (fchollet) [inline tensorflow/python/keras/layers/convolutional_test.py]: - Please add tests for the error messages added
- Please add a correctness test; current tests only check the output shape, not whether group convolution has actually taken place.

**Review by fchollet — CHANGES_REQUESTED** [review]
Thank you for the PR! Sorry for the delay in reviewing, I had missed this PR (I don't get email notifications from GitHub).

File: tensorflow/python/keras/layers/convolutional_test.py
Comment (lgeiger) [reply tensorflow/python/keras/layers/convolutional_test.py]: > Please add tests for the error messages added

:+1: I added tests for this in the latest commits

> Please add a correctness test; current tests only check the output shape, not whether group convolution has actually taken place.

I looked at the tests for Keras convolutional layers and the `tf.nn` and none of them test the actual correctness of the output. The testing for correctness of the grouped convolution and normal convolutions is already handled on the C++.
I can add a very simple test that checks that this implementation matches a simple loop based version of grouped convolutions. Do you think that would be useful?

**Review by lgeiger — COMMENTED** [review]
Thanks for the review, especially the error message are a lot better now :+1: 
Do you think it would be possible to cherry-pick these change onto the 2.2 release branch, or is that to late?

> Sorry for the delay in reviewing, I had missed this PR. (I don't get email notifications from GitHub)

Very sensible 📧 

I know you are working on moving the Keras project into it's own repo again which will greatly improve the contributor experience, but in the mean time what is the best way for outside contributors to raise attention to PRs that require a review from Keras team? (It would be great to get a review for #37552 as well)

File: tensorflow/python/keras/layers/convolutional_test.py
Comment (lgeiger) [reply tensorflow/python/keras/layers/convolutional_test.py]: > Please add a correctness test; current tests only check the output shape, not whether group convolution has actually taken place.

I added more tests in 2c703ebe0b3dd7ec0da4017be54cb15d268c5638

**Comment by byronyi** [comment]
Any updates? Grouped convolutions seem to be used a lot more these days and I look forward to seeing it in tf-keras.

**Comment by tnbalsam** [comment]
+1, this would be extremely useful to have.

**Comment by tanzhenyu** [comment]
Somehow this wasn't in my radar. Looking into it now.

**Comment by lgeiger** [comment]
> Somehow this wasn't in my radar. Looking into it now.

@tanzhenyu thanks for checking it out :+1:

File: tensorflow/python/keras/layers/convolutional.py
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/convolutional.py:142]: what if `self.groups = None`? Should we allow that and make None == 1?

File: tensorflow/python/keras/layers/convolutional_test.py
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/convolutional_test.py:290]: I think we are using the `with test_util.use_gpu():` these days

File: tensorflow/python/keras/layers/convolutional.py
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/convolutional.py:179]: I'm a little confused here, dont we need `self.groups` # of kernels each of shape `(kernel_size,) + (input_channel // self.groups, self.filters)`?

File: tensorflow/python/keras/layers/convolutional.py
Comment (lgeiger) [reply tensorflow/python/keras/layers/convolutional.py:142]: I don't think this case will come up very much in practice, but it is very easy to support. I added a quick check in d74951f42d8ae8b7c29fb428c3745c90196f95bc

File: tensorflow/python/keras/layers/convolutional_test.py
Comment (lgeiger) [reply tensorflow/python/keras/layers/convolutional_test.py]: :+1: I didn't know that, most of the tests in this file still use `cached_session`. Changed it in f3d9db564a7190c6c33e2c5f8747be631b1f6b8d

File: tensorflow/python/keras/layers/convolutional.py
Comment (lgeiger) [reply tensorflow/python/keras/layers/convolutional.py:179]: As far as I know most frameworks follow the convention that the shape of the weights is `(*kernel_size, input_channel // groups, output_channels)` so that `filters==output_channels` regardless of the number of groups used. E.g. PyTorch does the same [here](https://github.com/pytorch/pytorch/blob/82d58ed484eb04f894475de0551055f7e070f481/torch/nn/modules/conv.py#L47-L48).
I personally think of `filters` as the number of output channels of a convolution, so adopting the convention that `filters` are the number of channels per group doesn't seem natural to me, but please correct me if I am wrong, this seems to be more a question of definition.
For me figure 3 in https://arxiv.org/pdf/1611.05431.pdf suggests to prefer `filters==output_channels` as well.

**Review by lgeiger — COMMENTED** [review]
@tanzhenyu Thanks for the comments, I made the requested changes below.

File: tensorflow/python/keras/layers/convolutional.py
Comment (tanzhenyu) [reply tensorflow/python/keras/layers/convolutional.py:179]: Right, so `filters==output_channels`.
What I don't understand here is, let's say we want to map a (55, 55, 32) conv input to (27, 27, 64) conv output, with `groups=2`. So apparently we need 2 different kernels, both are with shape (3, 3, 16, 64). The first kernel will map the first half (55, 55, 16) to (27, 27, 64), and the second kernel will map the second half (55, 55, 16) to (27, 27, 64). And the final result would be the sum of the two.
So either we need 2 kernels in 1 conv layer, or we need 2 conv layers. Am I missing something here? (Feel free to point out where I'm wrong)

Comment (razorx89) [reply tensorflow/python/keras/layers/convolutional.py:179]: `filters` gets split into `groups` by the convolutional algorithm. If you would have multiple convolutions mapping to 64d, what would be the end result? 128d instead of 64d for `groups==2`? As far as I know this is handled by the convolution `tf.nn.conv*` or cudnn.

Comment (lgeiger) [reply tensorflow/python/keras/layers/convolutional.py:179]: @tanzhenyu The grouped convolution computes splits  in the output channel dimension which are then concatenated to combine the result. Your example above would change to:

The first kernel will map the first half (55, 55, 16) to (27, 27, ~~64~~**32**), and the second kernel will map the second half (55, 55, 16) to (27, 27, ~~64~~**32**). And the final result would be the ~~sum~~**concatenation** of the two.

The implementation of `tf.nn.conv2d` handles this computation either using XLA or CUDNN. This is the reason why adding grouped convolution support to Keras only requires a few additional lines that change the kernel shape. To be fair the fact that `tf.nn.conv2d` computes a grouped convolution depending on the shapes of the inputs and weights is a bit confusing in my opinion.

Consider the following low level TensorFlow implementation of your example that showcases the computation in relation to a simple loop based implementation and the relation of a grouped convolution to a depthwise:

```python
import numpy as np
import tensorflow as tf


@tf.function(experimental_compile=True)  # Needed for grouped conv CPU support
def conv_2d(inputs, weights, strides=[1, 1, 1, 1], padding="VALID"):
    return tf.nn.conv2d(inputs, weights, strides=strides, padding=padding)


# loop based convolution

groups = 2
inputs = tf.random.normal(shape=[1, 55, 55, 32], dtype=tf.float64)
weights = tf.random.normal(shape=[3, 3, 32 // groups, 64], dtype=tf.float64)

outputs = conv_2d(inputs, weights, strides=[1, 2, 2, 1])

grouped_inputs = tf.split(inputs, groups, axis=-1)
grouped_weights = tf.split(weights, groups, axis=-1)
outputs_ref = tf.concat(
    [
        conv_2d(inputs, weights, strides=[1, 2, 2, 1])
        for inputs, weights in zip(grouped_inputs, grouped_weights)
    ],
    axis=-1,
)

np.testing.assert_allclose(outputs.numpy(), outputs_ref.numpy())

# depthwise

inputs_dw = tf.random.normal(shape=[1, 55, 55, 64], dtype=tf.float64)
weights_dw = tf.random.normal(shape=[3, 3, 64, 1], dtype=tf.float64)
weights_grouped_dw = tf.reshape(weights_dw, [3, 3, 1, 64])

outputs_dw = conv_2d(inputs_dw, weights_grouped_dw, strides=[1, 2, 2, 1])
outputs_dw_ref = tf.nn.depthwise_conv2d(
    inputs_dw, weights_dw, strides=[1, 2, 2, 1], padding="VALID"
)

np.testing.assert_allclose(outputs_dw.numpy(), outputs_dw_ref.numpy())
```

Comment (tanzhenyu) [reply tensorflow/python/keras/layers/convolutional.py:179]: Ok I'm assuming this will end up calling tf.nn.conv2d with `inputs=(H,W, IN_CHANNEL)` and `filters=(IN_CHANNEL/GROUP, K, K, OUT_CHANNEL)` and under the hood (at kernel level) it will infer the group and get multiple OUT_CHANNEL/GROUP conv outputs and concatenate before returning.

**Review by tanzhenyu — DISMISSED** [review]
LGTM. (Though it'd be nice to add test for DepthwiseConv as well)

File: tensorflow/python/keras/layers/convolutional.py
Comment (tanzhenyu) [reply tensorflow/python/keras/layers/convolutional.py:179]: Oh I refreshed the page before your comment. Yeah this seems legit.

Comment (tanzhenyu) [reply tensorflow/python/keras/layers/convolutional.py:179]: @lgeiger Though to be more concise, I think when we call `tf.nn.conv2d`, the inputs don't need to be splitted, because we need to infer groups from input channel and kernel size?
(The docstring of that op needs to be updated as well)

Comment (lgeiger) [reply tensorflow/python/keras/layers/convolutional.py:179]: > Though to be more concise, I think when we call `tf.nn.conv2d`, the inputs don't need to be splitted, because we need to infer groups from input channel and kernel size?

That's correct and that's what this PR should do as well, since we don't actually change `Layer.call`.

**Comment by lgeiger** [comment]
Looks like `api_compatibility_test` fails on CI. I'll rebase and update the API goldens. This might take a few hours on my machine though since I don't have a warm build cache.

> Though it'd be nice to add test for DepthwiseConv as well

Good idea, I will do that once TF finished building on my machine. I didn't even think that we can use this for testing before writing https://github.com/tensorflow/tensorflow/pull/36773#discussion_r404294541

**Comment by tanzhenyu** [comment]
> Looks like `api_compatibility_test` fails on CI. I'll rebase and update the API goldens. This might take a few hours on my machine though since I don't have a warm build cache.
> 
> > Though it'd be nice to add test for DepthwiseConv as well
> 
> Good idea, I will do that once TF finished building on my machine. I only thought about that this can be used for testing when writing [#36773 (comment)](https://github.com/tensorflow/tensorflow/pull/36773#discussion_r404294541)

Yep I'm not sure if you can run `bazel run third_party/tensorflow/tools/api/tests:api_compatibility_test -- --update_goldens True`?

**Comment by lgeiger** [comment]
@tanzhenyu I improved the tests in 52160bc31c19a2fa1be5d8df3f09ccd50ea343d8 and rebased onto master so that the API snapshot update won't introduce any merge conflicts.

> Yep I'm not sure if you can run bazel run third_party/tensorflow/tools/api/tests:api_compatibility_test -- --update_goldens True

I am trying it now, but it's still compiling 😞. Feel free to update the API snapshots during the internal merge.

**Comment by tanzhenyu** [comment]
> @tanzhenyu I improved the tests in [52160bc](https://github.com/tensorflow/tensorflow/commit/52160bc31c19a2fa1be5d8df3f09ccd50ea343d8) and rebased onto master so that the API snapshot update won't introduce any merge conflicts.
> 
> > Yep I'm not sure if you can run bazel run third_party/tensorflow/tools/api/tests:api_compatibility_test -- --update_goldens True
> 
> I am trying it now, but it's still compiling 😞. Feel free to update the API snapshots during the internal merge.

There is no "internal merge" before this PR passes all CI builds (unfortunately). If the bazel run doesn't work, you can probably manually edit the .pbtxt files.

**Comment by lgeiger** [comment]
@tanzhenyu I updated the goldens in 212616f0cae39ac2ccfcc2988eca5fcb3161cdf4.

Unfortunately the `api_compatibility_test` segfaults on my machine, but it seems to work when only checking the `keras.layers` APIs endpoints, so hopefully CI will be happy with it.

Sorry for the cumbersome review process, but a fast response time for non-trivial changes is sadly not possible for me when the TF build takes an entire night to run 😉

**Review by tanzhenyu — APPROVED** [review]

**Comment by tanzhenyu** [comment]
> @tanzhenyu I updated the goldens in [212616f](https://github.com/tensorflow/tensorflow/commit/212616f0cae39ac2ccfcc2988eca5fcb3161cdf4).
> 
> Unfortunately the `api_compatibility_test` segfaults on my machine, but it seems to work when only checking the `keras.layers` APIs endpoints, so hopefully CI will be happy with it.
> 
> Sorry for the cumbersome review process, but a fast response time for non-trivial changes is sadly not possible for me when the TF build takes an entire night to run 😉

No worries. We're working on faster build time these days.
I'm slightly worried about the position of this argument. We would usually put this at end of all positional arguments for backward compatibility reasons.

**Comment by lgeiger** [comment]
> I'm slightly worried about the position of this argument. We would usually put this at end of all positional arguments for backward compatibility reasons.

Personally I think the current placement makes the most sense. It is before `activation` and `use_bias` that change the output after doing the convolution and before configuration of initializers and constraints. But it is the last argument that changes the actual convolution operations (i.e. after `filters`, `kernel_size`, `strides`, `dilation`, ...). This is also where [PyTorch places the groups argument](https://pytorch.org/docs/stable/nn.html#conv2d) though that shouldn't influence this decision too much.

I wouldn't expect most users to change the `data_format` format keyword, so my gut feeling is that every keyword argument after `data_format` will likely not be addressed via positional arguments anyway. But I agree technically this might be a breaking change for some users.
I am not sure what the general policy for TensorFlow is, but adding it behind the `name` keyword seems not very consistent. Are there any plans to make some arguments [keyword only](https://www.python.org/dev/peps/pep-3102/), so that future additions like this can be made without risking breaking user code?

**Comment by tanzhenyu** [comment]
> > I'm slightly worried about the position of this argument. We would usually put this at end of all positional arguments for backward compatibility reasons.
> 
> Personally I think the current placement makes the most sense. It is before `activation` and `use_bias` that change the output after doing the convolution and before configuration of initializers and constraints. But it is the last argument that changes the actual convolution operations (i.e. after `filters`, `kernel_size`, `strides`, `dilation`, ...). This is also where [PyTorch places the groups argument](https://pytorch.org/docs/stable/nn.html#conv2d) though that shouldn't influence this decision too much.
> 
> I wouldn't expect most users to change the `data_format` format keyword, so my gut feeling is that every keyword argument after `data_format` will likely not be addressed via positional arguments anyway. But I agree technically this might be a breaking change for some users.
> I am not sure what the general policy for TensorFlow is, but adding it behind the `name` keyword seems not very consistent. Are there any plans to make some arguments [keyword only](https://www.python.org/dev/peps/pep-3102/), so that future additions like this can be made without risking breaking user code?

Yeah putting `group` here makes sense. Though for compatibility reasons I would imagine putting it after activation would be better, since most users specify up until activation. (For Resnet they might set `use_bias=False` as well). But I believe they would most likely achieve this by keywords. 

AFAIK we don't have this plan. If you would like to contribute, that'd be great.

**Review by tanzhenyu — DISMISSED** [review]

**Comment by lgeiger** [comment]
> Though for compatibility reasons I would imagine putting it after activation would be better, since most users specify up until activation. (For Resnet they might set `use_bias=False` as well). But I believe they would most likely achieve this by keywords.

I cannot judge how important backwards compatibility to the TensorFlow team is so I'm happy to follow your lead. In my personal experience I had to make some adjustment after most minor TF version upgrades so as long as it is mentioned in the release notes I would prefer consistent argument placement over backwards compatibility in this case, but your mileage might vary.

> AFAIK we don't have this plan. If you would like to contribute, that'd be great.

I'm happy to contribute it, though this would definitely be a breaking change then (though one with a very clear error message and simple upgrade path).
This largely depends on if TensorFlow already accepts Python 3 only code, since AFAIK keyword only arguments are not supported in Python 2.

**Comment by byronyi** [comment]
> This largely depends on if TensorFlow already accepts Python 3 only code, since AFAIK keyword only arguments are not supported in Python 2.

@lgeiger The team only ships Python 3 binaries these days so I believe it's fine. Btw thanks for the detailed explanation of grouped/depthwise convolution in the review comments. Very helpful indeed.

**Comment by lgeiger** [comment]
@tanzhenyu Thanks for the approval! Looks like the tests are passing, is there anything that is left to do before this can be merged?

**Comment by dathudeptrai** [comment]
@lgeiger Hi, thanks for ur implementation. Did u test the speed ?, it seems that ur group conv very slow, even with group = input filters, it's still run very slow compared with normal convolution.

**Comment by lgeiger** [comment]
> Hi, thanks for ur implementation. Did u test the speed ?, it seems that ur group conv very slow, even with group = input filters, it's still run very slow compared with normal convolution.

I did not benchmark this code, as it only implements a user facing Keras API that relies on `tf.nn.convolution` to execute the actual grouped convolution either using CUDNN or XLA.

**Comment by lgeiger** [comment]
>  it seems that ur group conv very slow, even with group = input filters, it's still run very slow compared with normal convolution.

@dathudeptrai How does the performance of the native grouped convolutions implemented in this PR compare to a loop based implementation as mentioned in https://github.com/tensorflow/tensorflow/pull/36773#discussion_r404294541?

**Comment by dathudeptrai** [comment]
@lgeiger hi, the problem is not about this PR. The problem cause it run very slow is that i apply mixed precision training for group convolution (mixed preicision is still speedup when apply for normal convolution :D).

**Comment by lgeiger** [comment]
> hi, the problem is not about this PR. The problem cause it run very slow is that i apply mixed precision training for group convolution (mixed preicision is still speedup when apply for normal convolution :D).

Makes sense. @dathudeptrai I think it would be best to open a new issue for this.

**Comment by lgeiger** [comment]
@gbaned It looks like the PR is approved and CI, including the internal `import/copybara` review, is green. Is there a reason why this PR hasn't been merged, or is this a problem with copybara.

Let me know if there is still something that needs clarification.

File: tensorflow/python/keras/layers/convolutional.py
Comment (jsimsa) [inline tensorflow/python/keras/layers/convolutional.py:82]: (for tensorflow/api-owners)

The following seems to be copy pasted from PyTorch [documentation](https://pytorch.org/docs/stable/nn.html#torch.nn.Conv2d) and should be rephrased.

**Comment by gbaned** [comment]
@lgeiger Can you please check @jsimsa comments and keep us posted. Thanks!

File: tensorflow/python/keras/layers/convolutional.py
Comment (lgeiger) [reply tensorflow/python/keras/layers/convolutional.py:82]: The structure is heavily inspired by PyTorch because I thought it describes the `groups` argument quite clearly.
@jsimsa Coul you elaborate on what part of the docstring is unclear, or should I just rephrase it so it doesn't resemble the PyTorch version as closely?

File: tensorflow/python/keras/layers/convolutional.py
Comment (jsimsa) [inline tensorflow/python/keras/layers/convolutional.py:89]: should this say `filters / input_channels`?

File: tensorflow/python/keras/layers/convolutional.py
Comment (jsimsa) [reply tensorflow/python/keras/layers/convolutional.py:82]: You need to rephrase this.

File: tensorflow/python/keras/layers/convolutional.py
Comment (lgeiger) [reply tensorflow/python/keras/layers/convolutional.py]: Good catch :+1: Fixed in c87915e09e5d9db8eae7b017307fc4b22862acc9

File: tensorflow/python/keras/layers/convolutional.py
Comment (lgeiger) [reply tensorflow/python/keras/layers/convolutional.py]: @jsimsa I updated the docstrings in 389fba4b3c9337ae2930b605f8cae3891c8b60d7

I hope this makes them clearer and easier to understand.

**Review by jsimsa — APPROVED** [review]

**Comment by lgeiger** [comment]
CI currently fails on this PR due to a build failure that has been fixed on master, I am happy to rebase if that helps.

@gbaned @jsimsa @fchollet @tanzhenyu This PR has now been open for over two and a half months, is there anything that is still blocking this from being merged?

**Comment by byronyi** [comment]
This seems to be reverted in dd2ea875d92eeb83e81b1cb92e29e61d488e98b2.

**Comment by tanzhenyu** [comment]
> This seems to be reverted in [dd2ea87](https://github.com/tensorflow/tensorflow/commit/dd2ea875d92eeb83e81b1cb92e29e61d488e98b2).

It broke the gpu unit tests which was added in this PR.

Some obscured error message:
Traceback (most recent call last):
  File "/third_party/tensorflow/python/keras/layers/convolutional_test.py", line 393, in test_group_conv
    self.assertAllClose(layer(inputs), expected_outputs, rtol=1e-5)
  File "/third_party/tensorflow/python/framework/test_util.py", line 1217, in decorated
    return f(*args, **kwds)
  File "/third_party/tensorflow/python/framework/test_util.py", line 2598, in assertAllClose
    self._assertAllCloseRecursive(a, b, rtol=rtol, atol=atol, msg=msg)
  File "/build/work/01080b4d142eb76348ad5b7462050a4cb61f/google3/runfiles/google3/third_party/tensorflow/python/framework/test_util.py", line 2558, in _assertAllCloseRecursive
    (path_str, path_str, msg)))
  File "/third_party/tensorflow/python/framework/test_util.py", line 2451, in _assertArrayLikeAllClose
    a = self._GetNdArray(a)
  File "/build/work/01080b4d142eb76348ad5b7462050a4cb61f/google3/runfiles/google3/third_party/tensorflow/python/framework/test_util.py", line 2445, in _GetNdArray
    a = self.evaluate(a)
  File "/third_party/tensorflow/python/framework/test_util.py", line 2151, in evaluate
    return sess.run(tensors)
  File "/build/work/01080b4d142eb76348ad5b7462050a4cb61f/google3/runfiles/google3/third_party/tensorflow/python/framework/test_util.py", line 1683, in run
    return super(ErrorLoggingSession, self).run(*args, **kwargs)
  File "/third_party/tensorflow/python/client/session.py", line 958, in run
    run_metadata_ptr)
  File "/third_party/tensorflow/python/client/session.py", line 1181, in _run
    feed_dict_tensor, options, run_metadata)
  File "/third_party/tensorflow/python/client/session.py", line 1359, in _do_run
    run_metadata)
  File "/third_party/tensorflow/python/client/session.py", line 1384, in _do_call
    raise type(e)(node_def, op, message)
third_party.tensorflow.python.framework.errors_impl.FailedPreconditionError: 2 root error(s) found.
  (0) Failed precondition: Read of uninitialized variable 
	 [[node conv1d/BiasAdd/ReadVariableOp (defined at /third_party/tensorflow/python/keras/layers/convolutional_test.py:393) ]]
	 [[xla_compile]]
	 [[xla_run/_1]]
  (1) Failed precondition: Read of uninitialized variable 
	 [[node conv1d/BiasAdd/ReadVariableOp (defined at /third_party/tensorflow/python/keras/layers/convolutional_test.py:393) ]]
	 [[xla_compile]]

**Comment by lgeiger** [comment]
@tanzhenyu This is strange, since CI passed in this PR. Do you think this could be related to unrelated changes, since this PR has been opened almost three months ago?

Unfortunately I am currently not able to easily reproduce this since I don't have a GPU available. However, I checked that the added test passes when executing with XLA CPU.

Should I rebase this PR and open it again, so CI can run?

**Comment by lgeiger** [comment]
I resubmitted the changes in #39516

**Comment by lgeiger** [comment]
@gbaned @jsimsa @fchollet @tanzhenyu I resubmitted the changes at #39516 and locally verified that the tests don't fail when rebased on latest master. Could you please take another look and approve the resubmission so that CI can run?

This PR has now been pending for over 3 months and the RFC for even longer. It would be great to be able to land this soon.

**Comment by lgeiger** [comment]
#39516 has been merged, thank you all for the detailed reviews 🚀

**Pull request merged**
